src/mono/mono/mini/interp/jiterpreter.c

```diff
@@ -1491,6 +1545,105 @@ mono_jiterp_patch_opcode (volatile JiterpreterOpcode *ip, guint16 old_opcode, gu
 #endif
 }
 
+/*
+ * Unordered thread-local pointer queue (used for do_jit_call and interp_entry wrappers)
+ * The queues are all tracked in a global list so that it is possible to perform a global
+ *  'purge item with this value from all queues' operation, which means queue operations
+ *  are protected by a lock
+ */
+
+#define NUM_QUEUES 2
+pthread_key_t queue_keys[NUM_QUEUES] = { 0 };
+pthread_once_t queue_keys_initialized = PTHREAD_ONCE_INIT;
+mono_mutex_t queue_mutex;
+GPtrArray *shared_queues = NULL;
+
+static void
+free_queue (void *ptr) {
+	mono_os_mutex_lock (&queue_mutex);
+	g_ptr_array_remove_fast (shared_queues, ptr);
+	g_ptr_array_free ((GPtrArray *)ptr, TRUE);
+	mono_os_mutex_unlock (&queue_mutex);
+}
+
+static void
+initialize_queue_keys () {
+	mono_os_mutex_lock (&queue_mutex);
+	shared_queues = g_ptr_array_new ();
+	mono_os_mutex_unlock (&queue_mutex);
+
+	for (int i = 0; i < NUM_QUEUES; i++)
+		g_assert (pthread_key_create (&queue_keys[i], free_queue) == 0);
+}
+
+static pthread_key_t
+get_queue_key (int queue) {
+	g_assert ((queue >= 0) && (queue < NUM_QUEUES));
+	pthread_once (&queue_keys_initialized, initialize_queue_keys);
+	return queue_keys[queue];
+}
+
+static GPtrArray *
+get_queue (int queue) {
+	pthread_key_t key = get_queue_key (queue);
+	GPtrArray *result = NULL;
+	if ((result = (GPtrArray *)pthread_getspecific (key)) == NULL) {
+		pthread_setspecific (key, result = g_ptr_array_new ());
+		mono_os_mutex_lock (&queue_mutex);
+		g_ptr_array_add (shared_queues, result);
+		mono_os_mutex_unlock (&queue_mutex);
+	}
+	return result;
+}
+
+// Purges this item from all queues
+void
+mono_jiterp_tlqueue_purge_all (gpointer item) {
+	mono_os_mutex_lock (&queue_mutex);
+	for (int i = 0; i < shared_queues->len; i++) {
+		GPtrArray *queue = (GPtrArray *)g_ptr_array_index (shared_queues, i);
+		gboolean ok = g_ptr_array_remove_fast (queue, item);
+		if (ok) {
+			// g_printf ("Purged %x from queue %x\n", (unsigned int)item, (unsigned int)queue);
+		}
+	}
+	mono_os_mutex_unlock (&queue_mutex);
+}
+
+// Removes the next item from the queue, if any, and returns it (NULL if empty)
+EMSCRIPTEN_KEEPALIVE gpointer
+mono_jiterp_tlqueue_next (int queue) {
+	GPtrArray *items = get_queue (queue);
+	mono_os_mutex_lock (&queue_mutex);
+	gpointer result = NULL;
+	if (items->len) {
+		result = g_ptr_array_index (items, 0);
+		g_ptr_array_remove_index_fast (items, 0);
+	}
+	mono_os_mutex_unlock (&queue_mutex);
+	return result;
+}
+
+// Adds a new item to the end of the queue and returns the new size of the queue
+EMSCRIPTEN_KEEPALIVE int
+mono_jiterp_tlqueue_add (int queue, gpointer item) {
```

Comment:
These two methods are a bit dangerous/fragile (but you may want to keep them this way, anyway):

1. They don't do coop GC transitions on entry, which means they could potentially run while the runtime is in a GC pause.
2. They use a mono_os_mutex instead of a mono_coop_mutex which means that if the thread suspends while the lock is taken, it may deadlock the GC.

Both of these things are not a _problem_ right now because there is nothing interesting happening while the lock is taken (no runtime functions are called - just some unmanaged data structure manipulation).  But if that were ever to change...

Reply:
I can switch to using coop mutexes, that sounds worth doing.

Reply:
Well there are downsides: 
1. if this stuff can be called before/after mono's threading machinery is running, or after a thread is detached from mono but before it completely disappears, those coop primitives might not work right.
2. there'a. bit of stack copying that happens when we do `MONO_ENTER_GC_UNSAFE` (ie when these EMSCRIPTEN_KEEPALIVE functions are called) and when we do `MONO_ENTER_GC_SAFE` (ie when coop mutexes  try to take a lock). So there will be a perf hit.

If this needs to be fast, and you're confident that you will never do any runtime work in these three functions, it might just be enough to add a comment/warning.

Reply:
The before/after scenario is possible since these are called from setTimeout handlers, so I probably have to use mono_os_mutex.